lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java

```diff
@@ -1221,6 +1229,111 @@ private Location createListingLocation(String path)
         return createLocation(path);
     }
 
+    @Test
+    void testFileDoesNotExistUntilClosed()
+    {
+        assertThatNoException().isThrownBy(() -> {
+            try {
+                Location location = getRootLocation().appendPath("testFile2");
+                getFileSystem().deleteFile(location);
+                try (OutputStream out = getFileSystem().newOutputFile(location).create()) {
+                    assertFalse(fileExists(location));
+                    out.write("test".getBytes(UTF_8));
+                    assertFalse(fileExists(location));
+                }
+                assertTrue(fileExists(location));
+            }
+            catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+    }
+
+    @Test
+    @Disabled("This test takes 5 minutes to run locally")
+    public void testLargeFileDoesNotExistUntilClosed()
+    {
+        assertThatNoException().isThrownBy(() -> {
+            try {
+                Location location = getRootLocation().appendPath("testFile3");
+                getFileSystem().deleteFile(location);
+                byte[] bytes = new byte[8192];
+                Slice slice = Slices.wrappedBuffer(bytes);
+                for (int i = 0; i < bytes.length / 4; i++) {
+                    slice.setInt(i * 4, i);
+                }
+                try (OutputStream outputStream = getFileSystem().newOutputFile(location).create()) {
+                    // write a 32 mb file
+                    for (int i = 0; i < 32 * 1024 * 1024 / bytes.length; i++) {
```

Comment:
the intention here is to write exactly 32b, so it's important that division is exact

```
[IntMath.]divide(32 * 1024 * 1024 , bytes.length, [RoundingMode.]UNNECESSARY)
```